baseplate/frameworks/queue_consumer/kafka.py

```diff
@@ -0,0 +1,448 @@
+            # let this exception crash the server so we'll stop processing messages
+            # and won't commit offsets. when the server restarts it will get
+            # this message again and try to process it.
+            logger.exception(
+                "Unhandled error while trying to process a message, terminating the server"
+            )
+            raise
+
+
+class _BaseKafkaQueueConsumerFactory(QueueConsumerFactory):
+    def __init__(
+        self,
+        name: str,
+        baseplate: Baseplate,
+        bootstrap_servers: str,
+        group_id: str,
+        topics: Sequence[str],
+        handler_fn: Handler,
+        message_unpack_fn: Callable[[bytes], Dict[str, Any]] = json.loads,
+        health_check_fn: Optional[HealthcheckCallback] = None,
+    ):
+        self.baseplate = baseplate
+
+        service_name, _, group_name = group_id.partition(".")
+        assert service_name and group_name, "group_id must start with 'SERVICENAME.'"
+        assert name == f"kafka_consumer.{group_name}"
+
+        self.consumer = self._make_consumer(bootstrap_servers, group_id, topics)
+        self.name = f"kafka_consumer.{group_name}"
+
+        self.handler_fn = handler_fn
+        self.message_unpack_fn = message_unpack_fn
+        self.health_check_fn = health_check_fn
+
+    @classmethod
+    def _consumer_config(cls) -> Dict[str, Any]:
+        raise NotImplementedError
+
+    @classmethod
+    def _make_consumer(
+        cls, bootstrap_servers: str, group_id: str, topics: Sequence[str]
+    ) -> confluent_kafka.Consumer:
+        consumer_config = {
+            "bootstrap.servers": bootstrap_servers,
+            "group.id": group_id,
+            # reset the offset to the latest offset when no stored offset exists.
+            # this means that when a new consumer group is created it will only
+            # process new messages.
+            "auto.offset.reset": "latest",
+        }
+        consumer_config.update(cls._consumer_config())
+        consumer = confluent_kafka.Consumer(consumer_config)
+
+        try:
+            # we can allow a blocking timeout here because there is only one
+            # consumer for the entire server
+            metadata = consumer.list_topics(timeout=10)
+        except confluent_kafka.KafkaException:
+            logger.error("failed getting metadata from %s, exiting.", bootstrap_servers)
+            raise
+
+        all_topics = set(metadata.topics.keys())
+        for topic in topics:
+            assert (
+                topic in all_topics
+            ), f"topic '{topic}' does not exist. maybe it's misspelled or on a different kafka cluster?"
+
+        # pylint: disable=unused-argument
+        def log_assign(
+            consumer: confluent_kafka.Consumer, partitions: List[confluent_kafka.TopicPartition]
+        ) -> None:
+            for topic_partition in partitions:
+                logger.info("assigned %s/%s", topic_partition.topic, topic_partition.partition)
```

Comment:
:no_bell: I wish we could just let you use f-strings in `logging` calls :shakes-fist:

Reply:
yeah, that made me sad to have to change them